### ISSUE TITLE
software architecture diagram

### ISSUE THREAD

**Issue by russelljjarvis** [russelljjarvis/neuronunit#26]
Hi @rgerkin,

[hyperlinked](https://github.com/russelljjarvis/neuronunit/blob/dev/docs/bulletin_board.png) is a better version of the ASCII art diagram I emailed you. It could be improved and used for developer documentation. The same diagram is also available for editing at [this link](https://www.draw.io/#Hrusselljjarvis%2Finformatics_poster%2Fmaster%2Fbulletin_board.xml).

I have massively reduced the import statements in [nsga_parallel.py](https://github.com/russelljjarvis/neuronunit/blob/dev/neuronunit/optimization/nsga_parallel.py) and recovered a lot of the functionality of the code before we ventured to make `model` picklable, and also before `tempfile` in jNEUROML backend.

I realized that that it was no longer necessary to make files unique to CPUs such as in the line below:
`new_file_path = str(get_neab.LEMS_MODEL_PATH)+str(int(os.getpid()))`

if all file writes were eliminated. Unnecessary file writes where coming from the interaction between the tempfile/tempdir calls in jNEUROML backend, and parallel implementation. After I commented out those lines, I could also safely comment out the line above.

I have added these lines to the end of the Dockerfile in `neuronunit-optimization`
```
COPY stdout_worker.py $HOME/stdout_worker.py
COPY func2rc.sh $HOME/func2rc.sh
RUN cat $HOME/func2rc.sh >> ~/.bashrc
RUN sudo chown -R jovyan /home/jovyan/work
ENV QT_QPA_PLATFORM offscreen
ENV PYTHONPATH $PYTHONPATH/home/jovyan/neuronunit/neuronunit/optimization
RUN /bin/bash -c "source $HOME/func2rc.sh"
WORKDIR $HOME/neuronunit/neuronunit
```

My personal dockerbuilds could be broken because of a problem with `cmake`. This problem could be transient, so rather than patching it I just built on top of my existing docker images, in the docker stack. 

The code referenced by `COPY stdout_worker.py $HOME/stdout_worker.py` is the code you gave me for observing stdout on the clients.

The code referenced with `COPY func2rc.sh $HOME/func2rc.sh` is:

func2rc.sh: 

```
function sc(){
    export PYTHONPATH=$1
    ipcluster start -n 8 --profile=default &
    sleep 5
    python stdout_worker.py &
    ipython -i nsga_parallel.py
    # explanation
    # there is some ugly inconsistencies between the
    # controller and the engines on ipyparallel
    # basically the controllers fixed location of neuronunit is fixed to /home/jovyan/work/neuronunit/neuronunit
    # this probably
    # because PYTHONPATH was set to this early with ENV
    # this is very static and it can't be updated
    #ln -s "`$1`../../neuronunit  /home/jovyan/work/neuronunit/neuronunit"
}
```

I also deleted a lot of files do not correspond to active ideas/development, in an effort to boost the signal to noise ratio of the repositories code as a whole.

Additionally I have mounted neuronunit on the docker container, in a way such that it over writes the destination`$HOME/neuronunit/neuronunit`, as that seemed to be the default PYTHONPATH, that the engines drew from, and it seemed simpler to overwrite that than to edit sys.path.

**Comment by russelljjarvis** [comment]
I have created a condensed solution that is a corollary to the above code snippets.
This Dockerfile has an ENTRYPOINT that executes the latest dev source code. Its not perfect, but its close to the level of pre poster functionality, and it has some merges from scidash.
https://github.com/russelljjarvis/ParallelPyNEURON/blob/dev/neuronunit-optimization/Dockerfile

**Comment by rgerkin** [comment]
I've added a copy of this "Bulletin Board" mockup to scidash/assets/mockups, where I would like to keep other useful drawings like this.
